main/src/ca/uwaterloo/flix/api/lsp/provider/completion/CompletionUtils.scala

```diff
@@ -181,14 +182,49 @@ object CompletionUtils {
     def isInternal(decl: TypedAst.Def): Boolean = decl.spec.ann.isInternal
 
     val isPublic = decl.spec.mod.isPublic && !isInternal(decl)
-    val isMatch = decl.sym.text.startsWith(word)
+    val isMatch = looseMatch(word, decl.sym.text)
 
     isMatch && isPublic
   }
 
   /**
-   * Checks if we should offer AutoUseCompletion or AutoImportCompletion.
-   * Currently, we will only offer them if at least three characters have been typed.
-   */
+    * Checks if we should offer AutoUseCompletion or AutoImportCompletion.
+    * Currently, we will only offer them if at least three characters have been typed.
+    */
   def shouldComplete(word: String): Boolean = word.length >= 3
+
+  /**
+    * Returns `true` if the query is a loose match for the key.
+    * After splitting query and key by camel case, every query segment must be a prefix of some key segment in order.
+    *
+    * Example:
+    *   - looseMatch("fBT",  "fooBarTest") = true
+    *   - looseMatch("fBrT", "fooBarTest") = false
+    *   - looseMatch("fTB",  "fooBarTest") = false
+    *
+    * @param query  The query string, usually from the user input.
+    * @param key    The key string, usually from the completion item.
+    */
+  def looseMatch(query: String, key: String): Boolean = {
+    @tailrec
+    def matchSegments(query: List[String], key: List[String]): Boolean = (query, key) match {
+      case (Nil, _) => true
+      case (_, Nil) => false
+      case (qHead :: qTail, kHead :: kTail) =>
+        if (kHead.startsWith(qHead))
+          matchSegments(qTail, kTail)
+        else
+          matchSegments(query, kTail)
+    }
+    matchSegments(splitByCamelCase(query), splitByCamelCase(key))
+  }
+
+  /**
+    * Splits a string by camel case.
+    *
+    * Example: "fooBarTest" -> List("foo", "Bar", "Test")
+    */
+  private def splitByCamelCase(input: String): List[String] = {
+    input.split("(?<!^)(?=[A-Z])").toList
```

Comment:
Yeah this is not very common.
Link to the doc https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html
We are actually using these:
![QQ_1733222696474](https://github.com/user-attachments/assets/2f343ae5-93ee-4ea5-8da1-4f688763011e)
Since the `split` in Scala is not inclusive, we can not just match on the characters.

I removed the `(?<!^)`. It can prevent the leading "" in some cases, but here just `(?=[A-Z])` is fine.